lib/model/Product.dart

```diff
@@ -313,47 +314,86 @@ class Product extends JsonObject {
+      } else if (key
+          .startsWith(ProductField.CATEGORIES_TAGS_IN_LANGUAGES.key)) {
+        OpenFoodFactsLanguage lang =
+            _langFrom(key, ProductField.CATEGORIES_TAGS_IN_LANGUAGES.key);
         final values = _jsonValueToList(json[key]);
         if (lang != OpenFoodFactsLanguage.UNDEFINED && values != null) {
           result.categoriesTagsInLanguages ??= {};
           result.categoriesTagsInLanguages![lang] = values;
         }
-      } else if (key.startsWith('ingredients_tags_')) {
-        OpenFoodFactsLanguage lang = _langFrom(key, 'ingredients_tags_');
+      } else if (key
+          .startsWith(ProductField.INGREDIENTS_TAGS_IN_LANGUAGES.key)) {
+        OpenFoodFactsLanguage lang =
+            _langFrom(key, ProductField.INGREDIENTS_TAGS_IN_LANGUAGES.key);
         final values = _jsonValueToList(json[key]);
         if (lang != OpenFoodFactsLanguage.UNDEFINED && values != null) {
           result.ingredientsTagsInLanguages ??= {};
           result.ingredientsTagsInLanguages![lang] = values;
         }
-      } else if (key.startsWith('labels_tags_')) {
-        OpenFoodFactsLanguage lang = _langFrom(key, 'labels_tags_');
+      } else if (key.startsWith(ProductField.LABELS_TAGS_IN_LANGUAGES.key)) {
+        OpenFoodFactsLanguage lang =
+            _langFrom(key, ProductField.LABELS_TAGS_IN_LANGUAGES.key);
         final values = _jsonValueToList(json[key]);
         if (lang != OpenFoodFactsLanguage.UNDEFINED && values != null) {
           result.labelsTagsInLanguages ??= {};
           result.labelsTagsInLanguages![lang] = values;
         }
-      } else if (key.startsWith('countries_tags_')) {
-        OpenFoodFactsLanguage lang = _langFrom(key, 'countries_tags_');
+      } else if (key.startsWith(ProductField.COUNTRIES_TAGS_IN_LANGUAGES.key)) {
+        OpenFoodFactsLanguage lang =
+            _langFrom(key, ProductField.COUNTRIES_TAGS_IN_LANGUAGES.key);
         final values = _jsonValueToList(json[key]);
         if (lang != OpenFoodFactsLanguage.UNDEFINED && values != null) {
           result.countriesTagsInLanguages ??= {};
           result.countriesTagsInLanguages![lang] = values;
         }
-      } else if (key.startsWith('ingredients_text_')) {
-        OpenFoodFactsLanguage lang = _langFrom(key, 'ingredients_text_');
+      } else if (key
+          .startsWith(ProductField.INGREDIENTS_TEXT_IN_LANGUAGES.key)) {
+        OpenFoodFactsLanguage lang =
+            _langFrom(key, ProductField.INGREDIENTS_TEXT_IN_LANGUAGES.key);
         if (lang != OpenFoodFactsLanguage.UNDEFINED) {
           result.ingredientsTextInLanguages ??= {};
           result.ingredientsTextInLanguages![lang] = json[key];
         }
       }
+      // The OFF API also can give values in all available languages for some
+      // of the fields. We store those values in same kinds of fields as
+      // the product_name_[2 letter language code] values.
+      if (key == ProductField.NAME_ALL_LANGUAGES.key) {
```

Comment:
I think there's potential confusion here. Let me explain.
```dart
ProductField.NAME_IN_LANGUAGES: 'product_name_',
ProductField.NAME_ALL_LANGUAGES: 'product_name_languages',
// ...

if (key.startsWith(ProductField.NAME_IN_LANGUAGES.key)) {
// 'product_name_languages' matches, but is not a decent option
if (key == ProductField.NAME_ALL_LANGUAGES.key) {
// 'product_name_languages' does match AND is not a decent option
```

There are many `if else`s but your added code starts with a simple `if`, and after the `startsWith` tests.
In the case of `'product_name_languages'` a first `if` will be triggered BUT will always be irrelevant, and then "your" `if` will work.

My suggestion (in that file): always code with `else`, and start with "your" `== ProductField.NAME_ALL_LANGUAGES.key` test.

Something like that:
```dart
if (key == ProductField.NAME_ALL_LANGUAGES.key) {
} else if (key == ProductField.INGREDIENTS_TEXT_ALL_LANGUAGES.key) {
} else if (key.startsWith(ProductField.NAME_IN_LANGUAGES.key)) {
} else if (key.startsWith(ProductField.CATEGORIES_TAGS_IN_LANGUAGES.key)) {
} else if (key.startsWith(ProductField.INGREDIENTS_TAGS_IN_LANGUAGES.key)) {
} else if (key.startsWith(ProductField.LABELS_TAGS_IN_LANGUAGES.key)) {
} else if (key.startsWith(ProductField.COUNTRIES_TAGS_IN_LANGUAGES.key)) {
} else if (key.startsWith(ProductField.INGREDIENTS_TEXT_IN_LANGUAGES.key)) {
}
```

Reply:
Done